nav2_costmap_2d/plugins/plugin_container_layer.cpp

```diff
@@ -0,0 +1,235 @@
+// Copyright (c) 2024 Polymath Robotics, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "nav2_costmap_2d/plugin_container_layer.hpp"
+
+#include "nav2_costmap_2d/costmap_math.hpp"
+#include "nav2_costmap_2d/footprint.hpp"
+#include "nav2_util/node_utils.hpp"
+#include "rclcpp/parameter_events_filter.hpp"
+#include "pluginlib/class_list_macros.hpp"
+
+PLUGINLIB_EXPORT_CLASS(nav2_costmap_2d::PluginContainerLayer, nav2_costmap_2d::Layer)
+
+using std::vector;
+
+namespace nav2_costmap_2d
+{
+PluginContainerLayer::PluginContainerLayer()
+{
+}
+
+PluginContainerLayer::~PluginContainerLayer()
+{
+}
+
+void PluginContainerLayer::onInitialize()
+{
+  auto node = node_.lock();
+
+  if (!node) {
+    throw std::runtime_error{"Failed to lock node"};
+  }
+
+  nav2_util::declare_parameter_if_not_declared(node, name_ + "." + "enabled",
+      rclcpp::ParameterValue(true));
+  nav2_util::declare_parameter_if_not_declared(node, name_ + "." + "plugins",
+      rclcpp::ParameterValue(std::vector<std::string>{}));
+  nav2_util::declare_parameter_if_not_declared(node, name_ + "." + "combination_method",
+      rclcpp::ParameterValue(1));
+
+  node->get_parameter(name_ + "." + "enabled", enabled_);
+  node->get_parameter(name_ + "." + "plugins", plugin_names_);
+
+  int combination_method_param{};
+  node->get_parameter(name_ + "." + "combination_method", combination_method_param);
+  combination_method_ = combination_method_from_int(combination_method_param);
+
+  dyn_params_handler_ = node->add_on_set_parameters_callback(
+    std::bind(
+      &PluginContainerLayer::dynamicParametersCallback,
+      this,
+      std::placeholders::_1));
+
+
+  plugin_types_.resize(plugin_names_.size());
+
+  for (size_t i = 0; i < plugin_names_.size(); ++i) {
+    plugin_types_[i] = nav2_util::get_plugin_type_param(node, name_ + "." + plugin_names_[i]);
+  }
+
+  // Then load and add the plug-ins to the costmap
+  for (unsigned int i = 0; i < plugin_names_.size(); ++i) {
+    std::shared_ptr<Layer> plugin = plugin_loader_.createSharedInstance(plugin_types_[i]);
+    addPlugin(plugin, plugin_names_[i]);
+  }
+
+  default_value_ = nav2_costmap_2d::FREE_SPACE;
```

Comment:
Unused?

Reply:
required for initial matchSize() in initialization, was undefined and populating map with non zero values after initial matchSize()

Reply:
Should you use the master costmap's value for this? Or, I figure actually the safer thing to do is set it as unknown space, so that the costmap internal layers themselves can set it to FREE if they choose. The free should override unknown, but the "FREE" here could override otherwise what is intended to be tracked as UNKNOWN -- right?

Reply:
That makes considerably more sense, fixed in latest commit